rust/storage/src/network_admission_control.rs

```diff
@@ -0,0 +1,146 @@
+use super::{GetError, Storage};
+use chroma_error::{ChromaError, ErrorCodes};
+use futures::{future::Shared, FutureExt, StreamExt};
+use parking_lot::Mutex;
+use std::{collections::HashMap, future::Future, pin::Pin, sync::Arc};
+use thiserror::Error;
+use tracing::{Instrument, Span};
+
+#[derive(Clone)]
+pub struct NetworkAdmissionControl {
+    storage: Storage,
+    outstanding_requests: Arc<
+        Mutex<
+            HashMap<
+                String,
+                Shared<
+                    Pin<
+                        Box<
+                            dyn Future<Output = Result<(), Box<NetworkAdmissionControlError>>>
+                                + Send
+                                + 'static,
+                        >,
+                    >,
+                >,
+            >,
+        >,
+    >,
+}
+
+#[derive(Error, Debug, Clone)]
+pub enum NetworkAdmissionControlError {
+    #[error("Error performing a get call from storage {0}")]
+    StorageGetError(#[from] GetError),
+    #[error("IO Error")]
+    IOError,
+    #[error("Error deserializing to block")]
+    DeserializationError,
+}
+
+impl ChromaError for NetworkAdmissionControlError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            NetworkAdmissionControlError::StorageGetError(e) => e.code(),
+            NetworkAdmissionControlError::IOError => ErrorCodes::Internal,
+            NetworkAdmissionControlError::DeserializationError => ErrorCodes::Internal,
+        }
+    }
+}
+
+impl NetworkAdmissionControl {
+    pub fn new(storage: Storage) -> Self {
+        Self {
+            storage,
+            outstanding_requests: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    pub async fn read_from_storage<F, R>(
```

Comment:
does this need to be pub? Seems like get() is the public api right? maybe im confused